source/common/router/router.cc

```diff
@@ -687,14 +687,18 @@ Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_strea
   // a backoff timer.
   ASSERT(upstream_requests_.size() <= 1);
 
-  bool buffering = (retry_state_ && retry_state_->enabled()) || !active_shadow_policies_.empty();
+  bool buffering =
+      (retry_state_ && retry_state_->enabled()) || !active_shadow_policies_.empty() ||
+      (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.internal_redirects_with_body") &&
```

Comment:
Still thinking about how to cleanly re-use `enableInternalRedirectsWithBody()` from the HCM - my thinking here is that there are multiple levels where we want to check this (HCM, FM, Router), HCM has the longest lifetime so it should be what determines whether this runtime guard is enabled or not. This is currently just a convenience to make progress on tests while I think of a way to access this value from the Router filter.

I think checking this from objects with two different lifetimes could put Envoy in a weird state until they converge, where short-lived items (e.g. Router) try to do internal redirects with bodies before HCM has updated.

Reply:
I think this is fine - we just won't redirect unless the flag is true both when the router filter is created and when the hcm is created right?

Reply:
Yep that's right. The only possible concern I think would be switching from off->on, the router will start buffering requests before the HCM starts actually doing internal redirects.